Write logs to disk in packaged builds

jsantell opened the pull request:
Fixed #1448.

Lots of wrestling with webpack here. Required for working on things that are packaged builds only, like #1418.

Ping @Mossop for review, and ping @victorporof for if there's a better way to handle the webpack build stuff to not include main process stuff in app/shared/logging.js in the renderer process. I don't want to bikeshed here, but this was a lot of work to get webpack to even build, silence linting and fixing dependency tests for this file loaded in all environments.

I can now see logs in real time in OSX's Console in ~/Library/Logs/Tofino/tofino.log -- curious if it's as good in Windows/Linux too

jsantell commented on app/main/log-utils.js:
```diff
@@ -0,0 +1,86 @@
+ * Logic and default paths for platforms inspired by electron-log
+ * https://github.com/megahertz/electron-log
+ */
+export const getLogDirectory = function({
```
This was originally async for fs access, but then the logger would have to buffer input as it's included in all environments synchronously. Should be async in the future so we're not doing this at startup, fine for now IMO

jsantell commented on build/config/webpack.base.js:
```diff
@@ -31,7 +31,8 @@ export default {
     alias: {
-      'dtrace-provider': path.join(Const.BUILD_SYSTEM_DIR, 'config', 'empty_module.js'),
+      'dtrace-provider': 'empty-module',
+      'tofino-log-utils': 'empty-module',
```
Using the same line as dtrace-provider for tofino-log-utils was breaking the build. Not sure why.

Mossop replied:
Maybe because that file doesn't exist anymore for some reason. Maybe I forgot to git add it.

victorporof replied:
Bad path, should be "utils", not "config". Not sure how I missed this during the initial review.

jsantell commented on build/config/webpack.main.js:
```diff
@@ -33,6 +33,11 @@ export default {
   ],
+  resolve: {
+    alias: {
+      'tofino-log-utils': path.join(SRC_DIR, 'log-utils.js'),
```
Appended 'tofino-' to the module just as an indication it may be aliased. Flexible with other idioms if you have any ideas

Mossop replied:
👍

jsantell commented:
Windows build failing loading build-config, guessing during when the logger is loaded during the build phase, and just use console.log. I think we shouldn't attempt to use the logger for building (as build-config won't exist). Thoughts?

```
Error: Cannot find module '../build-config'
    at Function.Module._resolveFilename (module.js:438:15)
    at Function.Module._load (module.js:386:25)
    at Module.require (module.js:466:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (C:/projects/tofino-u1hv8/app/shared/logging.js:19:1)
    at Module._compile (module.js:541:32)
    at loader (C:\projects\tofino-u1hv8\node_modules\babel-register\lib\node.js:144:5)
    at Object.require.extensions.(anonymous function) [as .js] (C:\projects\tofino-u1hv8\node_modules\babel-register\lib\node.js:154:7)
    at Module.load (module.js:456:32)
    at tryModuleLoad (module.js:415:12)
```

jsantell commented:
Oh, and during build, I get more dtrace spam but doesn't seem to have any negative effects. What webpack magic do we need to just stop dtrace screaming at us?

```
build INFO: Executing /Users/jsantell/Dev/tofino/.electron/Electron.app/Contents/MacOS/Electron
{ Error: Cannot find module "./build"
    at webpackMissingModule (webpack:///./~/dtrace-provider/dtrace-provider.js?:17:67)
    at eval (webpack:///./~/dtrace-provider/dtrace-provider.js?:17:151)
    at Object.<anonymous> (/Users/jsantell/Dev/tofino/lib/main/index.js:122:2)
    at __webpack_require__ (/Users/jsantell/Dev/tofino/lib/main/index.js:20:30)
    at eval (webpack:///./~/bunyan/lib/bunyan.js?:79:18)
    at Object.<anonymous> (/Users/jsantell/Dev/tofino/lib/main/index.js:95:2)
    at __webpack_require__ (/Users/jsantell/Dev/tofino/lib/main/index.js:20:30)
    at eval (webpack:///./app/shared/logging.js?:61:45)
    at Object.<anonymous> (/Users/jsantell/Dev/tofino/lib/main/index.js:86:2)
    at __webpack_require__ (/Users/jsantell/Dev/tofino/lib/main/index.js:20:30) code: 'MODULE_NOT_FOUND' }
{ Error: Cannot find module "./build"
    at webpackMissingModule (webpack:///./~/dtrace-provider/dtrace-provider.js?:17:67)
    at eval (webpack:///./~/dtrace-provider/dtrace-provider.js?:17:151)
    at Object.<anonymous> (/Users/jsantell/Dev/tofino/lib/main/index.js:122:2)
    at __webpack_require__ (/Users/jsantell/Dev/tofino/lib/main/index.js:20:30)
    at eval (webpack:///./~/bunyan/lib/bunyan.js?:79:18)
    at Object.<anonymous> (/Users/jsantell/Dev/tofino/lib/main/index.js:95:2)
    at __webpack_require__ (/Users/jsantell/Dev/tofino/lib/main/index.js:20:30)
    at eval (webpack:///./app/shared/logging.js?:61:45)
    at Object.<anonymous> (/Users/jsantell/Dev/tofino/lib/main/index.js:86:2)
    at __webpack_require__ (/Users/jsantell/Dev/tofino/lib/main/index.js:20:30) code: 'MODULE_NOT_FOUND' }
{ Error: Cannot find module "./build"
    at webpackMissingModule (webpack:///./~/dtrace-provider/dtrace-provider.js?:17:67)
    at eval (webpack:///./~/dtrace-provider/dtrace-provider.js?:17:151)
    at Object.<anonymous> (/Users/jsantell/Dev/tofino/lib/main/index.js:122:2)
    at __webpack_require__ (/Users/jsantell/Dev/tofino/lib/main/index.js:20:30)
    at eval (webpack:///./~/bunyan/lib/bunyan.js?:79:18)
    at Object.<anonymous> (/Users/jsantell/Dev/tofino/lib/main/index.js:95:2)
    at __webpack_require__ (/Users/jsantell/Dev/tofino/lib/main/index.js:20:30)
    at eval (webpack:///./app/shared/logging.js?:61:45)
    at Object.<anonymous> (/Users/jsantell/Dev/tofino/lib/main/index.js:86:2)
    at __webpack_require__ (/Users/jsantell/Dev/tofino/lib/main/index.js:20:30) code: 'MODULE_NOT_FOUND' }
```

Mossop commented:
Is there a way we can use the profile directory rather than making up a new one?

jsantell commented:
We could -- I could look into more about common log areas, no preference on where

victorporof commented on app/main/log-utils.js:
```diff
@@ -0,0 +1,86 @@
+/**
+ * This file is loaded in app/shared/logging via `tofino-log-utils` alias,
+ * since we don't want to include node APIs in the renderer process.
+ */
```
😢

victorporof commented on app/main/log-utils.js:
```diff
@@ -0,0 +1,86 @@
+    [env['XDG_CONFIG_HOME']],
+    [env['HOME'], '.config'],
+    [env['XDG_DATA_HOME']],
+    [env['HOME'], '.local', '.share'],
```
Do we really want to allow so much variability here? Why not just log everything to a file in the profile directory? Just import environment.js. See how session.json is created.

victorporof commented on app/main/log-utils.js:
```diff
@@ -0,0 +1,86 @@
+  ],
+};
+
+function getFirstAvailableDirectory(dirs, appName) {
```
Ugh. `profile/log.txt` would be so much cleaner.

victorporof commented on app/shared/logging.js:
```diff
@@ -90,7 +94,30 @@ switch (name) {
+          period: '1d',
+          count: 3,
+        });
+      }
```
It'd be cleaner to split all of this out in a standalone function.

victorporof commented on build/utils/base-config.js:
```diff
@@ -14,4 +14,8 @@ export default {
+
+  // The `packaged` flag indicates whether or not the build
+  // is a distributed, standalone exexcutable or not.
+  packaged: false,
```
👍

victorporof commented on test/unit/lint/test-dependencies.js:
```diff
@@ -69,6 +69,10 @@ describe('dependencies', () => {
       'stream',
+      // Used for bunyan logging.
+      'path',
+      // Used for bunyan logging.
```
Probably not necessary to duplicate this comment 3 times.

jsantell commented:
Removed empty-module dependency with correct pathing, fixes the build issue with dtrace, addressed other comments, sans log directory to use profile directory

jsantell commented:
What's left? Add a try/catch around build-config so CI won't fail (still think the build step should not use our shared logger IMO)? And can change to use the profile dir.

jsantell commented:
Updated to handle build-config while building and use the profile directory
